R/QFeatures-class.R

```diff
@@ -667,36 +690,268 @@ longFormat <- function(object,
+                     assayLinks) {
+    ## Check arguments
+    stopifnot(inherits(x, "QFeatures"))
+    y <- .checkAssaysToInsert(y, x, name, replace = FALSE)
+    
+    ## Check (or create) assayLinks
+    if (!missing(assayLinks)) {
+        if (inherits(assayLinks, "AssayLink"))
+            assayLinks <- AssayLinks(assayLinks)
+        if (!identical(sort(names(assayLinks)), sort(names(y))))
+            stop("'assayLinks' must be named after the assay(s) in 'y'.")
+    } else {
+        assayLinks <- AssayLinks(names = names(y))
+    }
+    
+    ## Update the colData
+    cd <- .updateColData(x, y)
+    
+    ## Add the assay to the ExperimentList
+    ## NOTE: we replace using the `@` slot. Although not recommended, 
+    ## this bypasses the checks of all the elements (using 
+    ## `validObject`) in the ExperimentList as this is already 
+    ## performed when building the QFeatures object and `y` is checked
+    ## at the beginning of the function. This leads to a reduction in
+    ## computational time. 
+    el <- experiments(x)
+    for(ii in names(y)) {
+        el@listData[[ii]] <- y[[ii]]
+    }
+    
+    ## Update the sampleMap
+    smap <- MultiAssayExperiment:::.sampleMapFromData(cd, el)
+    
+    ## Update the AssayLinks
+    al <- append(x@assayLinks, assayLinks)
+    
+    ## Update the QFeatures object with all the updated parts
+    BiocGenerics:::replaceSlots(
+        object = x,
+        ExperimentList = el,
+        colData = cd,
+        sampleMap = smap,
+        assayLinks = al,
+        check = FALSE
+    )
+}
+
+##' @md
+##'
+##' @rdname QFeatures-class
+##'
+##' @export
+removeAssay <- function(x, i) {
+    if (is.numeric(i) || is.logical(i))
+        i <- names(x)[i]
+    x[, , !names(x) %in% i]
+}
+
+##' @md
+##'
+##' @rdname QFeatures-class
+##'
+##' @export
+replaceAssay <- function(x, 
+                         y,
+                         i) {
+    ## Check arguments
     stopifnot(inherits(x, "QFeatures"))
-    el0 <- x@ExperimentList@listData
-    if (is.list(y)) el1 <- y
-    else el1 <- structure(list(y), .Names = name[1])
-    el <- ExperimentList(c(el0, el1))
-    smap <- MultiAssayExperiment:::.sampleMapFromData(colData(x), el)
-    if (inherits(assayLinks, "AssayLink"))
-        assayLinks <- AssayLinks(assayLinks)
-    new("QFeatures",
+    if (!missing(i) && (is.numeric(i) || is.logical(i)))
+        i <- names(x)[i]
+    y <- .checkAssaysToInsert(y, x, i, replace = TRUE)
+    
+    ## Update the colData
+    cd <- .updateColData(x, y)
+    
+    ## Replace the assay to the ExperimentList
+    ## NOTE: we replace using the `@` slot. Although not recommended, 
+    ## this bypasses the checks of all the elements (using 
+    ## `validObject`) in the ExperimentList as this is already 
+    ## performed when building the QFeatures object. This leads to a 
+    ## reduction in computational time. 
+    el <- experiments(x)
+    for(ii in names(y)) {
+        el@listData[[ii]] <- y[[ii]]
```

Comment:
Should we consider checking the validity of the new assays?

Reply:
Yes, this is performed at line 783